src/grt/grt-wave_options-parse.adb

```diff
@@ -0,0 +1,317 @@
+--  GHDL Run Time (GRT) -  command line options.
+--  Copyright (C) 2002 - 2014 Tristan Gingold
+--
+--  GHDL is free software; you can redistribute it and/or modify it under
+--  the terms of the GNU General Public License as published by the Free
+--  Software Foundation; either version 2, or (at your option) any later
+--  version.
+--
+--  GHDL is distributed in the hope that it will be useful, but WITHOUT ANY
+--  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+--  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+--  for more details.
+--
+--  You should have received a copy of the GNU General Public License
+--  along with GCC; see the file COPYING.  If not, write to the Free
+--  Software Foundation, 59 Temple Place - Suite 330, Boston, MA
+--  02111-1307, USA.
+--
+--  As a special exception, if other files instantiate generics from this
+--  unit, or you link this unit with other files to produce an executable,
+--  this unit does not by itself cause the resulting executable to be
+--  covered by the GNU General Public License. This exception does not
+--  however invalidate any other reasons why the executable file might be
+--  covered by the GNU Public License.
+
+with System; use System;
+with Grt.Types; use Grt.Types;
+with Grt.Strings; use Grt.Strings;
+with Grt.Vstrings; use Grt.Vstrings;
+with Grt.Errors; use Grt.Errors;
+
+with Grt.Wave_Options.Parse.Debug;
+
+package body Grt.Wave_Options.Parse is
+
+   procedure Start (Option_File : String) is
+      Stream : FILEs := File_Open (Option_File);
+      First, Last : Integer;
+      Line : String (1 .. Buf_Size);
+      Lineno : Natural;
+   begin
+      File_Path := new String'(Option_File);
+      Lineno := 0;
+
+      loop -- Reads line after line
+         exit when fgets (Line'Address, Line'Length, Stream) = Null_Address;
+         Lineno := Lineno + 1;
+
+         Last := New_Line_Pos (Line) - 1;
+         if Last < 0 then
+            Last := Line'Last;
+         end if;
+
+         First := First_Non_Whitespace_Pos (Line (Line'First .. Last));
+         if First = -1 or else Line (First) = '#' then
+            goto Continue;
+         end if;
+
+         Last := Last_Non_Whitespace_Pos (Line (First .. Last));
+         declare
+            Line_Str : String (1 .. Last + 1 - First) := Line (First .. Last);
+         begin
+            Line_Context :=
+               new Line_Context_Type'(Str => new String'(Line_Str),
```

Comment:
On 23/07/16 09:23, Jonsba wrote:
> In src/grt/grt-wave_options-parse.adb
> <https://github.com/tgingold/ghdl/pull/121#discussion_r71968896>:
>
>> +         Last := New_Line_Pos (Line) - 1;
>> +         if Last < 0 then
>> +            Last := Line'Last;
>> +         end if;
>> +
>> +         First := First_Non_Whitespace_Pos (Line (Line'First .. Last));
>> +         if First = -1 or else Line (First) = '#' then
>> +            goto Continue;
>> +         end if;
>> +
>> +         Last := Last_Non_Whitespace_Pos (Line (First .. Last));
>> +         declare
>> +            Line_Str : String (1 .. Last + 1 - First) := Line (First .. Last);
>> +         begin
>> +            Line_Context :=
>> +               new Line_Context_Type'(Str => new String'(Line_Str),
>
> I first write it that way, but there is a subtility :
>
> If the line begin with one (or more) space, First will be 2 (or higher)
> and then the index of the first character of Line_Context.Str will be 2
> (or higher) instead of 1.

Right.  But then you could also write:
   subtype line_str_type is string (1 .. last + 1 - first);
...
   new Line_Str_Type'(Line (First .. Last));

But:
> Then I had an issue in the Check_If_Found procedure in
> grt-wave_options-read.adb with the function call to Find, but now I just
> found out that if I change the line :
> Index := 1;

You should avoid to hard-code bounds!

> to :
> Index := Cursor.Line_Context.Str'First;
> then the issue is fixed.

Yes!